knowageutils/src/main/java/it/eng/spagobi/utilities/rest/RestUtilities.java

```diff
@@ -493,4 +495,44 @@ public static Map<String, String> toHeaders(HttpServletRequest request) {
 		return headers;
 	}
 
+	private static void addHeadersToRequest(Entry<String, String> entry, HttpMethodBase method) {
+		String headerValue = entry.getValue();
+		String headerKey = entry.getKey();
+		headerValue = RestUtilities.sanitizeValueFromRequestHeader(headerValue);
+		RestUtilities.checkIfValueFromRequestHeaderIsInWhitelistFromPropertiesFile(headerKey);
+		method.addRequestHeader(headerKey, headerValue);
+	}
+
+	private static String sanitizeValueFromRequestHeader(String value) {
+		return value.replaceAll("\r", "").replaceAll("%0d", "").replaceAll("%0D", "").replaceAll("\n", "").replaceAll("%0a", "").replaceAll("%0A", "");
+	}
+
+	private static void checkIfValueFromRequestHeaderIsInWhitelistFromPropertiesFile(String value) {
+		List<String> whitelist = RestUtilities.getWhitelistFromPropertiesFile();
+		if (!whitelist.contains(value))
+			throw new SpagoBIRuntimeException("Header value " + value + " is not in the list of allowed headers.");
+	}
+
+	private static List<String> getWhitelistFromPropertiesFile() {
+		Properties properties = new Properties();
+		List<String> whitelist = new ArrayList<>();
+
+		try (InputStream inputStream = RestUtilities.class.getClassLoader().getResourceAsStream("headers.properties")) {
```

Comment:
I don't think it's useful to read the file every time: this file needs a build to be changed. Please create a singleton class which cache the properties in memory.

Reply:
Fixed. I added WhitelistCache class, it will check if the whitelist is empty before trying to read the config file (file shouldn't be empty).